nix/tahoe-lafs.nix

```diff
@@ -70,6 +70,24 @@ buildPythonPackage rec {
     };
   };
 
+  postPatch =
+    let
+      versionFileContents = version: ''
+        # This _version.py is generated by tahoe-lafs.nix.
+        # TODO: We can have more metadata after we switch to flakes.
+        # Then the `self` input will have a `sourceInfo` attribute telling
+        __pkgname__ = "tahoe-lafs"
+        real_version = "${version}"
+        full_version = "${version}"
+        branch = ""
+        verstr = "${version}"
+        __version__ = verstr
+      '';
+    in
+      ''
+        cp ${builtins.toFile "_version.py" (versionFileContents version)} src/allmydata/_version.py
```

Comment:
Probably don't need this with hatch?

Reply:
Not sure this will be the only thing to change to finalize the switch to hatch.

Reply:
The build succeeds w/o this postPAtch, but then the test fails.

@hacklschorsch: this block you've added earlier here is no longer working on top of the new master for a reason I do not comprehend yet.

It apparently fails to write the (destination?) file for some obscure permissions (see those [logs](https://app.circleci.com/pipelines/github/LeastAuthority/tahoe-lafs/1234/workflows/8fe91bdb-b456-412f-b7f3-944f37ef8932/jobs/15083) for instance):

```
python3.12-tahoe-lafs> * Building wheel...
python3.12-tahoe-lafs> Traceback (most recent call last):
python3.12-tahoe-lafs>   File "/nix/store/h7gkm0drrmlx3r5qhfj4y4055bx8nd17-python3.12-pyproject-hooks-1.2.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
python3.12-tahoe-lafs>     main()
python3.12-tahoe-lafs>   File "/nix/store/h7gkm0drrmlx3r5qhfj4y4055bx8nd17-python3.12-pyproject-hooks-1.2.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 373, in main
python3.12-tahoe-lafs>     json_out["return_val"] = hook(**hook_input["kwargs"])
python3.12-tahoe-lafs>                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
python3.12-tahoe-lafs>   File "/nix/store/h7gkm0drrmlx3r5qhfj4y4055bx8nd17-python3.12-pyproject-hooks-1.2.0/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 280, in build_wheel
python3.12-tahoe-lafs>     return _build_backend().build_wheel(
python3.12-tahoe-lafs>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
python3.12-tahoe-lafs>   File "/nix/store/xvi3zndl6rdc77f9ppj05hngnrwcamwl-python3.12-hatchling-1.25.0/lib/python3.12/site-packages/hatchling/build.py", line 58, in build_wheel
python3.12-tahoe-lafs>     return os.path.basename(next(builder.build(directory=wheel_directory, versions=['standard'])))
python3.12-tahoe-lafs>                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
python3.12-tahoe-lafs>   File "/nix/store/xvi3zndl6rdc77f9ppj05hngnrwcamwl-python3.12-hatchling-1.25.0/lib/python3.12/site-packages/hatchling/builders/plugin/interface.py", line 147, in build
python3.12-tahoe-lafs>     build_hook.initialize(version, build_data)
python3.12-tahoe-lafs>   File "/nix/store/n8w58cihzwhw1gbj1vzn6gr42m1n5kv5-python3.12-hatch-vcs-0.4.0/lib/python3.12/site-packages/hatch_vcs/build_hook.py", line 35, in initialize
python3.12-tahoe-lafs>     dump_version(self.root, self.metadata.version, self.config_version_file, **kwargs)
python3.12-tahoe-lafs>   File "/nix/store/lflc8h97ld3p0ximi4qhrdlrr2317pgf-python3.12-setuptools-scm-8.1.0/lib/python3.12/site-packages/setuptools_scm/_integration/dump_version.py", line 58, in dump_version
python3.12-tahoe-lafs>     write_version_to_path(
python3.12-tahoe-lafs>   File "/nix/store/lflc8h97ld3p0ximi4qhrdlrr2317pgf-python3.12-setuptools-scm-8.1.0/lib/python3.12/site-packages/setuptools_scm/_integration/dump_version.py", line 94, in write_version_to_path
python3.12-tahoe-lafs>     target.write_text(content, encoding="utf-8")
python3.12-tahoe-lafs>   File "/nix/store/zv1kaq7f1q20x62kbjv6pfjygw5jmwl6-python3-3.12.7/lib/python3.12/pathlib.py", line 1047, in write_text
python3.12-tahoe-lafs>     with self.open(mode='w', encoding=encoding, errors=errors, newline=newline) as f:
python3.12-tahoe-lafs>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
python3.12-tahoe-lafs>   File "/nix/store/zv1kaq7f1q20x62kbjv6pfjygw5jmwl6-python3-3.12.7/lib/python3.12/pathlib.py", line 1013, in open
python3.12-tahoe-lafs>     return io.open(self, mode, buffering, encoding, errors, newline)
python3.12-tahoe-lafs>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
python3.12-tahoe-lafs> PermissionError: [Errno 13] Permission denied: '/tmp/nix-build-python3.12-tahoe-lafs-1.19.0.post1.drv-0/source/src/allmydata/_version.py'
python3.12-tahoe-lafs> ERROR Backend subprocess exited when trying to invoke build_wheel
```

Locally on my nixos, it tries a different path, but still fails:

```
PermissionError: [Errno 13] Permission denied: '/build/source/src/allmydata/_version.py'
```

Any idea?

Reply:
I'm puzzled as for what have we changed since our successful builds that could have impacted the path and/or the permissions...

Reply:
Why are you trying to manually write this file?

Reply:
> Why are you trying to manually write this file?

@hacklschorsch has implemented this here in cf82b0b.

And I assume that it is needed so `allmydata.test.test_client.Basic.test_versions` can `PASS`.

This test is the only one to `FAIL` when I remove this `postPatch` section (`version` is then `unknown` and should not be).

Reply:
That's a generated file; why can't Nix generate it the same way everything else does (instead of manually writing it)?

Reply:
I'm not a Nix developer, but I suppose they would say: "immutability". Or something like that.
For instance, the content of this generated file will always be the same given the same build inputs, even the timestamp will be the same all the time.

Anyway: I've found a way to make the `build` happy with this `_version.py`, just had to `chmod 0644 src/allmydata/_version.py` :relaxed:.

But the test still fails:

```
Traceback (most recent call last):                                                                                                                                           
  File "/nix/store/q6sn76xs45lxvhadz2hl1ks4x3y1xjcn-python312-tahoe-lafs/lib/python3.12/site-packages/twisted/internet/defer.py", line 2014, in _inlineCallbacks
    result = context.run(gen.send, result)                                                                                                                                   
  File "/home/user/workspace/Tahoe-LAFS/tahoe-lafs/src/allmydata/test/test_client.py", line 602, in test_versions
    self.failIfEqual(str(allmydata.__version__), "unknown")                                                                                                                  
  File "/nix/store/q6sn76xs45lxvhadz2hl1ks4x3y1xjcn-python312-tahoe-lafs/lib/python3.12/site-packages/twisted/trial/_synctest.py", line 491, in assertNotEqual
    raise self.failureException(msg or f"{first!r} == {second!r}")                                                                                                           
twisted.trial.unittest.FailTest: 'unknown' == 'unknown'
```

I'll dig further tomorrow. I need to find where this version is computed and why it won't take this file as input as it did before.

@meejah: maybe you already know? Was this `_version.py` trick and setup.py feature we no longer have?

Reply:
Previously, `setup.py` had a custom command to compute and (re-)write the `_version.py` file. Now, `hatch-vcs` via `hatch` should be doing that.

That said, the test is failing because the version ends up being "unknown" so something in the Nix stuff appears to be failing to find the right thing? (Or something like that).

You can make `hatch` re-generate the file via a couple  different commands, so doing custom interpolation of the file shouldn't be necessary. That is, even if you need to "pre-generate" the file for Nix purposes, you should be able to make `hatch` do it for you.